src/main/scala/org/apache/sysml/api/dl/CaffeLayer.scala

```diff
@@ -0,0 +1,343 @@
+// Fuse 
+class BatchNorm(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer with HasWeight with HasBias {
+  // val scale =  
+  override def sourceFileName = "spatial_batch_norm"
+  override def init(dmlScript:StringBuilder) = invokeInit(dmlScript, List[String](gamma, beta, ema_mean, ema_var), numChannels)
+  var update_mean_var = true
+  def forward(dmlScript: StringBuilder, isPrediction: Boolean): Unit = {
+    val mode = if(isPrediction) "\"test\"" else "\"train\""
+    invokeForward(dmlScript, List[String](out, withSuffix(ema_mean), withSuffix(ema_var), withSuffix(cache_mean), withSuffix(cache_var), withSuffix(cache_norm)), 
+        X, gamma, beta, numChannels, Hin, Win, mode, ema_mean, ema_var,  ma_fraction, eps)  
+  }
+  
+  def backward(dmlScript: StringBuilder, outSuffix:String): Unit = {
+    invokeBackward(dmlScript, outSuffix, List[String](dX, dgamma, dbeta), dout, out, ema_mean, ema_var, cache_mean, cache_var, cache_norm, X, gamma, beta, numChannels, 
+          Hin, Win, "\"train\"", ema_mean, ema_var,  ma_fraction, eps)
+  }
+  
+  private def withSuffix(str:String):String = if(update_mean_var) str else str + "_ignore"
+  override def weight = "ema_mean" + id
+  override def bias = "ema_var" + id
+  def cache_mean(): String = "cache_mean" + id
+  def cache_var():String = "cache_mean" + id
+  def cache_norm():String = "cache_norm" + id
+  var scaleLayer:Scale = null
+  def gamma():String = { checkNextLayer(); scaleLayer.weight }
+  def ma_fraction():String = if(param.getBatchNormParam.hasMovingAverageFraction()) param.getBatchNormParam.getMovingAverageFraction.toString else "0.999"
+  def eps():String = if(param.getBatchNormParam.hasEps()) param.getBatchNormParam.getEps.toString else "1e-5"
+  def beta():String = { checkNextLayer(); scaleLayer.bias }
+  def dgamma():String = { checkNextLayer();  scaleLayer.dWeight }
+  def dbeta():String = { checkNextLayer();  scaleLayer.dBias }
+  override def shouldUpdateWeight():Boolean = false
+  override def shouldUpdateBias():Boolean = false
+  def ema_mean(): String = weight
+  def ema_var(): String = bias
+  def checkNextLayer(): Unit = {
+    if(scaleLayer == null) {
+      val topLayers = net.getTopLayers(param.getName).map(l => net.getCaffeLayer(l)).toList
+      if(topLayers.length != 1 && !topLayers(0).isInstanceOf[Scale]) throw new LanguageException("Only one top layer of type Scale allowed for BatchNorm")
+      scaleLayer = topLayers(0).asInstanceOf[Scale]
+    }
+  }
+  def numChannels = bottomLayerOutputShape._1
+  def Hin = bottomLayerOutputShape._2
+  def Win = bottomLayerOutputShape._3
+}
+// weight is gamma and bias is beta
+class Scale(val param:LayerParameter, val id:Int, val net:CaffeNetwork) extends CaffeLayer with HasWeight with HasBias {
+  if(!param.getScaleParam.getBiasTerm) throw new LanguageException("Add \"scale_param { bias_term: true }\" to the layer " + param.getName)
+  override def sourceFileName = null
```

Comment:
Yes, I was planning to change that. Since I had recently did major refactoring, I wanted to stage the code-base.